signing_service/src/signing_service.py

```diff
@@ -0,0 +1,151 @@
+import socket
+from time import sleep
+
+from constants import SIGNING_SERVICE_DEFAULT_PORT
+from constants import SIGNING_SERVICE_MAXIMUM_RECONNECT_TIME
+
+
+logger = logging.getLogger()
+crash_logger = logging.getLogger('crash')
+
+
+class SigningService:
+    """
+    The Signing Service connects to Middleman as a client but then listens for requests coming from Concent via
+    Middleman. The underlying protocol is TCP and data sent over that is expected to conform to the Wire protocol.
+    """
+
+    def __init__(self, host, port, reconnect_delay) -> None:
+        assert isinstance(host, str)
+        assert isinstance(port, int)
+        assert isinstance(reconnect_delay, int)
+        self.host = host
+        self.port = port
+        self.reconnect_delay = reconnect_delay
+        self.current_reconnect_delay = 0
+        self.socket: socket.socket = None  # type: ignore
+        self.is_sigterm_sent = False
+
+    def run(self) -> None:
+        """
+        Handles main connection loop and its exceptions.
+
+        If connection is interrupted due to a failure, it should try to reconnect.
+        If connection is closed by other side, exit gracefully.
+        If a shutdown signal or KeyboardInterrupt is caught, exit gracefully.
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+
+        def _set_is_sigterm_sent_true(signum, frame):  # pylint: disable=unused-argument
+            self.is_sigterm_sent = True
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, _set_is_sigterm_sent_true)
+
+        while True:
+            if self.is_sigterm_sent:
+                self.socket.close()
+                logger.info('Closing connection and exiting on SIGTERM.')
+                break
+            try:
+                if self.socket is None:
+                    self._connect()
+                else:
+                    self.socket.send(b'')
+            except (socket.error, socket.gaierror) as exception:  # TODO: Test
+                logger.error(f'Socket error occurred: {exception}')
+
+                # When the other side closes the connection, return without errors
+                if isinstance(exception.args, tuple) and exception.args[0] == socket.errno.EPIPE:  # type: ignore
+                    self.socket.close()
+                    logger.info(os.strerror(exception.args[0]))
+                    break
+
+                # Increase delay and reconnect if the connection is interrupted due to a failure.
+                self._increase_delay()
+                self.socket = None
+            except KeyboardInterrupt:
+                # Handle keyboard interrupt.
+                self.socket.close()
+                logger.info('Closing connection and exiting on KeyboardInterrupt.')
+                break
+            except Exception as exception:
+                # Try to close opened socket.
+                self.socket.close()
```

Comment:
You're doing `socket.close()` in all the cases. I think you should have a `try ... finally` block that closes the connection instead. It would be less repetitive and the cleanup would be guaranteed. I think that for example now `SystemExit` (which does not inherit from `Exception` but from `BaseException`) will bypass your handlers and exit without closing the socket.

Reply:
:+1: done